libraries/Microsoft.Bot.Builder.Core/TurnContextServiceCollection.cs

```diff
@@ -0,0 +1,123 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Runtime.Serialization;
+using System.Text;
+
+namespace Microsoft.Bot.Builder
+{
+    /// <summary>
+    /// Represents a set of collection of services associated with the <see cref="ITurnContext"/>.
+    /// </summary>
+    /// <remarks>
+    /// TODO: add more details on what kind of services can/should be stored here, by whom and what the lifetime semantics are, etc.
+    /// </remarks>
+    public interface ITurnContextServiceCollection : IEnumerable<KeyValuePair<string, object>>
```

Comment:
>IEnumerable [](start = 53, length = 11)

Why not make this IDictionary<string, object>?  Don't I want to lookup by key?

Reply:
So nothing looks up by key now. The interface exposes `Get` as the primary way of accessing services and that's what 99% of callers would use. There are interesting cases however where being able to see which services are registered and maybe even filtering them by name or type is attractive, so for those scenarios supporting just `IEnumerable<KeyValuePair<string, object>>` prevents any other sort of mutability, but offers full inspection using something like LINQ APIs. In fact, this is what the extension method `TurnContextServiceCollectionExtensions::GetServices` does.

To give the best example of how this might be used, let's think about the auto loading of state we talked about. There could be _N_ `IStateManager`s registered as services (e.g. conversation, user, bot, whatever). My auto-load middleware can now do something like this to find all of those and tell them to load all their state in parallel:

```CSharp
var stateManagerServices = context.Services.GetServices<IStateManager>().ToList();

            if(EnableAutoLoad && stateManagerServices.Count > 0)
            {
                await Task.WhenAll(stateManagerServices.Select(sms => sms.Value.Load()));
            }
```

(NOTE: I am actually writing that thing right now and that is copy pasted from my initial implementation)